### PR TITLE
Do not render unless necessary

### PR DESCRIPTION
**Pull request by jasperSpeicher**
In [this PR](https://github.com/wavespotter/tell-tale/pull/1752) we found that extra renders in the mapbox context provider were contributing to infinite recursion when rendering the map. This PR cleans up memoization for the provider to only change the value when it is actually necessary.

This also fixes a crash that could happen if the map is loaded really skinny:
image

Tested locally with an RC package on npm.

### CODE REVIEW

File: src/components/MapboxMap/index.tsx [hunk 1]
```diff
@@ -78,13 +78,20 @@ const MapboxMap: React.FC<MapboxMapProps> = ({
...
-        bearing: newMap.getBearing(),
-        pitch: newMap.getPitch(),
-      });
+      const newCenter = newMap.getCenter();
+      const zoom = newMap.getZoom();
+      const center: [number, number] = [newCenter.lng, newCenter.lat];
+      const bearing = newMap.getBearing();
+      const pitch = newMap.getPitch();
+      // no need to update the state if the values are the same
+      if (zoom !== transform?.zoom && center !== transform?.center && bearing !== transform?.bearing && pitch !== transform?.pitch) {
```
Comment (aeveleth) [inline src/components/MapboxMap/index.tsx:87]: Doesn't `[0, 0] !== [0, 0]` evaluate to true? Probably best to use lodash's `isEqual` for the center comparison.

**Review by aeveleth — DISMISSED** [review]
Not sure what the check failures are, but I noticed one thing that is probably an issue.

File: src/components/MapboxMap/index.tsx (hunk 1 above)
Comment (jasperSpeicher) [reply src/components/MapboxMap/index.tsx:87]: Good call thank you

Comment (jasperSpeicher) [reply src/components/MapboxMap/index.tsx]: see e96c0b48913e44f7c6c92ec06b2123cb08cf332c

File: src/components/MapboxMap/index.tsx [hunk 2]
```diff
@@ -78,13 +81,24 @@ const MapboxMap: React.FC<MapboxMapProps> = ({
...
+      const zoom = newMap.getZoom();
+      const center: [number, number] = [newCenter.lng, newCenter.lat];
+      const bearing = newMap.getBearing();
+      const pitch = newMap.getPitch();
+      // no need to update the state if the values are the same
+      if (zoom !== transformRef.current?.zoom
+        && center[0] !== transformRef.current?.center[0]
+        && center[1] !== transformRef.current?.center[1]
+        && bearing !== transformRef.current?.bearing
+        && pitch !== transformRef.current?.pitch) {
```
[the remark this answers is not in the document]
Comment (jasperSpeicher) [reply src/components/MapboxMap/index.tsx]: Ugh. I have been experimenting with copilot and I missed this. I am turning it off.
See 71334ad8aa1b78a0d6af89a9c75ca2975ee54c29

File: src/components/MapboxMap/index.tsx [hunk 3]
```diff
@@ -52,12 +52,33 @@ const MapboxMap: React.FC<MapboxMapProps> = ({
...
+  // maintain this reference so we can access it in the useEffect below without including it in the dependency array
+  const transformRef = useRef(transform);
+  transformRef.current = transform;
 
   // Let the parent component overwrite the map bounds
   useDeepCompareEffectNoCheck(() => {
     if (!map || !fitBounds) return;
-    map.fitBounds(fitBounds.bounds, fitBounds.options);
-  }, [fitBounds]);
+    const { height, width } = map.getContainer().getBoundingClientRect();
```
Comment (juliusiv) [inline src/components/MapboxMap/index.tsx:62]: is this related to the renders?

**Review by juliusiv — COMMENTED** [review]
the PR description says "DO NOT MERGE", is that still the case?

**Comment by jasperSpeicher** [comment]
> the PR description says "DO NOT MERGE", is that still the case?

Sorry, removed that. It was fixed by the jotai PR

File: src/components/MapboxMap/index.tsx (hunk 3 above)
Comment (jasperSpeicher) [reply src/components/MapboxMap/index.tsx:62]: This is a separate bug I found while testing. I mentioned it in the description briefly.

**Review by juliusiv — APPROVED** [review]
